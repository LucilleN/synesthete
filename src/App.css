.App {
  /* margin-left: 5rem;
  margin-right: 5rem; */
  width: 100%;
  /* padding-left: 5rem;
  padding-right: 5rem; */
}
