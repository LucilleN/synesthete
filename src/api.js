let api = 'https://misconfigured-app.com/'
// const API_KEY = 'dc6zaTOxFJmzC' // Giphy's public beta key (thank you Giphy).
const authorizationUrl = 'https://accounts.spotify.com/api/token'
const authorizationKey = 'Basic MTU5YWM4OGIxYzUzNGVkN2FlNDE2MDJmMWU1NThhNDk6YmUxOTFmMTg3ZGZlNDgzMjg3MDAxZDNhNWZlYTEyNTM='
const authorizationHeaders = {
  'Content-Type': 'application/x-www-form-urlencoded',
  'Authorization': authorizationKey
}

const getAccessToken = () => {
  const response = fetch(authorizationUrl, {
    authorizationHeaders
  }).then(okCheck, emitNativeError)
    .then(response => response.json())

  return response.access_token
}

const apiHost = host => { api = host }
const urlFor = resource => `${api}${resource}`

const HTTP_OK = 200

const throwResponseError = response => {
  const error = new Error(response.statusText)
  error.response = response
  throw error
}

const emitNativeError = error => {
  throw error
}

const statusCheck = successStatuses => response => {
  if (successStatuses.includes(response.status)) {
    return response
  } else {
    throwResponseError(response)
  }
}

const okCheck = statusCheck([HTTP_OK])

const getHeaders = () => {
  const token = getAccessToken()
  const headers = {
    'Authorization': `Bearer ${token}`
  }
  return headers
}

//
// const paramsWithApiKey = params => {
//   const result = new URLSearchParams(params)
//   result.set('api_key', API_KEY)
//   return result
// }

// The fetch function initiates a connection to the web service.
// fetch returns a _promise_: an object that represents a future result.
// Thus, the function actually returns right away. However, when the
// anticipated result does show up, the code specifies what to do using
// either `then` or `catch`. Both functions accept another function,
// to be called upon a successful or failed promise, respectively.
// Furthermore, then `then` function can be chained: its return result
// is passed to the next `then` function as an argument. Here, the initial
// handler converts the raw result into JSON. That JSON then goes to the
// next handler, which does the actual work of putting the result on the
// web page.
//
// The design of fetch allows this entire sequence to be rendered in a
// _single statement_, thus obviating the need for curly braces but
// resulting in what many will view to be a decrease in readability
// (for those who arenâ€™t used to functional-style programming). YMMV

// const query = (resource, params) => fetch(`${urlFor(resource)}?${paramsWithApiKey(params)}`, {
const query = (resource, params) => {
  const optionalParams = (params) ? `?${new URLSearchParams(params)}` : ""

  const headers = getHeaders()

  fetch(`${urlFor(resource)}${optionalParams}`, {
    headers
  }).then(okCheck, emitNativeError)
    .then(response => response.json())
}

// const searchGifs = params => query('gifs/search', params)
const searchSongs = params => query('search', params)

const getRecommendation = params => query('recommendations', params)

const getAudioFeatures = songID => query(`audio-features/${songID}`)


export {
  apiHost,
  // searchGifs
  searchSongs,
  getAudioFeatures,
  getRecommendation
}
