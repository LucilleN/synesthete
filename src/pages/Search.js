import React, { useState } from 'react'
import Grid from '@material-ui/core/Grid'
import Typography from '@material-ui/core/Typography'
import { withStyles } from '@material-ui/core/styles'
import { Redirect } from 'react-router-dom';

import SearchResults from '../components/SearchResults'

import { searchSongs } from '../api'

const styles = theme => ({
  title: {
    color: theme.palette.light.pink,
    padding: 5,
    fontSize: '3rem',
    textAlign: 'center'
  },
  searchForm: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center'
  },
  searchField: {
    width: '50%',
    height: '1.5rem',
    marginBottom: 10
  },
  buttonEnabled: {
    color: theme.palette.white,
    fontSize: '1rem',
    background: theme.palette.dark.pink,
    height: '50px',
    width: '150px',
    borderRadius: '30px',
    border: 'none',
    margin: 5
  },
  buttonDisabled: {
    color: theme.palette.white,
    fontSize: '1rem',
    background: theme.palette.dark.pink,
    height: '50px',
    width: '150px',
    borderRadius: '30px',
    border: 'none',
    margin: 5,
    opacity: '0.3'
  }
})

const SearchForm = (props) => {

  const { classes } = props

  const [error, setError] = useState(null) // so it's const even though error can change?
  const [query, setQuery] = useState('')
  const [songs, setSongs] = useState([])
  const [songSelected, setSongSelected] = useState(false)
  const [songID, setSongID] = useState(null)

  const handleQueryChange = event => setQuery(event.target.value)

  const performQuery = async event => {
    event.preventDefault() // why?

    setError(null) // why?

    try {
      const result = await searchSongs({
        // rating: 'pg-13',
        q: query
      })

      setSongs(result.tracks.items)
    } catch (error) {
      setError('Sorry, but something went wrong.')
    }
  }

  const handleTrackSelect = (event) => {
    // do api stuff
    console.log(`CALLED HANDLE TRACK SELECT: selected track id ${event.target.id}`)
    setSongSelected(true)
    setSongID(event.target.id)
  }

  if (!songSelected) {

    return (
      <form className={classes.searchForm} onSubmit={performQuery}>
        <Typography className={classes.title}>
          search
        </Typography>

        <input name="query" type="text" value={query} onChange={handleQueryChange} className={classes.searchField}/>

        <div className="ButtonBar">
          <button type="submit" disabled={!query} className={query ? classes.buttonEnabled : classes.buttonDisabled}>
            search tracks
          </button>
        </div>

        {error && (
          <div className="error">
            {error}
          </div>
        )}

        <Grid container xs={10}>
          <SearchResults results={songs} onTrackSelect={handleTrackSelect}/>
        </Grid>

      </form>
    )
  }
  else {
    return (
      <Redirect to={`/visualization/${songID}`} />
    )
  }
}

export default withStyles(styles)(SearchForm)
