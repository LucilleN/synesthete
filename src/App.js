import React, { useEffect } from 'react'
import { Route, BrowserRouter as Router, Switch } from 'react-router-dom'

import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles'
import { withStyles } from '@material-ui/core/styles'

import Homepage from './pages/Homepage'
import About from './pages/About'
import Search from './pages/Search'
import UploadVisualization from './pages/UploadVisualization'
import Visualization from './pages/Visualization'
import MenuBar from './components/MenuBar'

import { apiHost } from './api'

export const theme = createMuiTheme({
  palette: {
    primary: {
      main: '#c23063', // magenta
      dark: '#180142' // dark purple
    },
    type: 'dark',
    dark: {
      pink: '#bf265c',
      purple: '#1c004f',
    },
    medium: {
      purple: '#5a0b66',
    },
    light: {
      pink: '#db91aa',
      purple: '#ac91db'
    },
    white: '#FFFFFF',
  },
  typography: {
    fontFamily: ['Dosis', 'Roboto', 'Open Sans', '"Helvetica Neue"', 'Arial', 'sans-serif'].join(',')
  }
})

export const defaultErrorText = 'Sorry, but something went wrong.'

const styles = () => ({
  root: {
    width: '100vw',
    height: '100%',
    minHeight: '100vh',
    background: `linear-gradient(0deg, ${theme.palette.dark.pink} 0%, ${theme.palette.dark.purple} 80%)`
  }
})

const routes = (
  <Router>
    <div>
      <MenuBar />
      <Switch>
        <Route exact path="/" component={Homepage} />
        <Route exact path="/about" component={About} />
        <Route exact path="/search" component={Search} />
        <Route exact path="/upload" component={UploadVisualization} />
        <Route exact path="/visualization" component={Visualization} />
        <Route exact path="/visualization/:songID" component={Visualization} />
      </Switch>
    </div>
  </Router>
)

const App = (props) => {

  useEffect(() => apiHost('https://api.spotify.com/v1/'))

  const { classes } = props

  return (
    <div className={classes.root}>
      <MuiThemeProvider theme={theme}>
        {routes}
      </MuiThemeProvider>
    </div>
  )
}

export default withStyles(styles)(App)
