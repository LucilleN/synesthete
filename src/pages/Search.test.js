import React from 'react'
import ReactDOM from 'react-dom'
import TestRenderer from 'react-test-renderer'
import ReactTestUtils from 'react-dom/test-utils'
import { MuiThemeProvider } from '@material-ui/core/styles'

import sinon from 'sinon'

import { theme } from '../App'
import Search from './Search'
import * as api from '../api'

let component;
beforeEach(() => {
  component = TestRenderer.create(
    <MuiThemeProvider theme={theme}>
      <Search />
    </MuiThemeProvider>
  )
})

class TestComponent extends React.Component {
  render() {
    return (
      <MuiThemeProvider theme={theme}>
        <Search />
      </MuiThemeProvider>
    )
  }
}

describe('initial state', () => {
  it('should start with a search field with default text', () => {
    // const component = TestRenderer.create(
    //   <MuiThemeProvider theme={theme}>
    //     <Search />
    //   </MuiThemeProvider>
    // )
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should start with a disabled search button', () => {
    // const component = TestRenderer.create(unstyledSearch)
    const tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

xdescribe('search button', () => {
  let div
  beforeEach(() => {
    div = document.createElement('div')
    ReactTestUtils.act(() => {
      ReactDOM.render(TestComponent, div)
    })
  })

  afterEach(() => ReactDOM.unmountComponentAtNode(div))

  it('should be enabled when the search field is not blank', () => {
    const searchInput = div.querySelector('input')
    ReactTestUtils.act(() => {
      searchInput.value = 'i can haz unit tests'
      ReactTestUtils.Simulate.change(searchInput)
    })

    const searchButton = div.querySelector('button')
    expect(searchButton.disabled).toBe(false)
  })

  it('should be disabled when the query is blank', () => {
    const searchButton = div.querySelector('button')
    expect(searchButton.disabled).toBe(true)
  })

  it('should be enabled when the user enters a query', () => {
    const searchInput = div.querySelector('input')
    ReactTestUtils.act(() => {
      searchInput.value = ''
      ReactTestUtils.Simulate.change(searchInput)
    })

    const searchButton = div.querySelector('button')
    expect(searchButton.disabled).toBe(false)
  })
})

// Helper function for the next two test collections.
const setupAndQuerySearchForm = async () => {
  const div = document.createElement('div')
  ReactTestUtils.act(() => {
    ReactDOM.render(unstyledSearch, div)
  })

  const searchInput = div.querySelector('input')
  ReactTestUtils.act(() => {
    searchInput.value = 'hello'
    ReactTestUtils.Simulate.change(searchInput)
  })

  const searchForm = div.querySelector('form')
  await ReactTestUtils.act(async () => {
    await ReactTestUtils.Simulate.submit(searchForm)
  })

  return div
}

describe('API calls', () => {
  let div
  beforeEach(async () => {
    sinon.stub(api, 'searchSongs')

    // To manage size, we supply a mock response that contains _only_ what the app will need. This does mean
    // that we need to revise the mock response if our app starts using more (or different) data.
    api.searchSongs.returns(Promise.resolve({
      "tracks" : {
        "href" : "https://api.spotify.com/v1/search?query=the+night+we+met&type=track&offset=0&limit=20",
        "items" : [ {
          "album" : {
            "album_type" : "album",
            "artists" : [ {
              "external_urls" : {
                "spotify" : "https://open.spotify.com/artist/6ltzsmQQbmdoHHbLZ4ZN25"
              },
              "href" : "https://api.spotify.com/v1/artists/6ltzsmQQbmdoHHbLZ4ZN25",
              "id" : "6ltzsmQQbmdoHHbLZ4ZN25",
              "name" : "Lord Huron",
              "type" : "artist",
              "uri" : "spotify:artist:6ltzsmQQbmdoHHbLZ4ZN25"
            } ],
            "available_markets" : [ "CA", "US" ],
            "external_urls" : {
              "spotify" : "https://open.spotify.com/album/3yoNZlqerJnsnMN5EDwwBS"
            },
            "href" : "https://api.spotify.com/v1/albums/3yoNZlqerJnsnMN5EDwwBS",
            "id" : "3yoNZlqerJnsnMN5EDwwBS",
            "images" : [ {
              "height" : 640,
              "url" : "https://i.scdn.co/image/d0a15f9288814a7e3f199ab2d1a6b0c9a9a633b5",
              "width" : 640
            }, {
              "height" : 300,
              "url" : "https://i.scdn.co/image/9bcea4ada214a061cecce6d565319f839ea83ce1",
              "width" : 300
            }, {
              "height" : 64,
              "url" : "https://i.scdn.co/image/0a2f11966d40cd6fc87ee70b71d8547578e87fe8",
              "width" : 64
            } ],
            "name" : "Strange Trails",
            "release_date" : "2015-04-07",
            "release_date_precision" : "day",
            "total_tracks" : 14,
            "type" : "album",
            "uri" : "spotify:album:3yoNZlqerJnsnMN5EDwwBS"
          },
          "artists" : [ {
            "external_urls" : {
              "spotify" : "https://open.spotify.com/artist/6ltzsmQQbmdoHHbLZ4ZN25"
            },
            "href" : "https://api.spotify.com/v1/artists/6ltzsmQQbmdoHHbLZ4ZN25",
            "id" : "6ltzsmQQbmdoHHbLZ4ZN25",
            "name" : "Lord Huron",
            "type" : "artist",
            "uri" : "spotify:artist:6ltzsmQQbmdoHHbLZ4ZN25"
          } ],
          "available_markets" : [ "CA", "US" ],
          "disc_number" : 1,
          "duration_ms" : 208211,
          "explicit" : false,
          "external_ids" : {
            "isrc" : "US53Q1200103"
          },
          "external_urls" : {
            "spotify" : "https://open.spotify.com/track/0QZ5yyl6B6utIWkxeBDxQN"
          },
          "href" : "https://api.spotify.com/v1/tracks/0QZ5yyl6B6utIWkxeBDxQN",
          "id" : "0QZ5yyl6B6utIWkxeBDxQN",
          "is_local" : false,
          "name" : "The Night We Met",
          "popularity" : 76,
          "preview_url" : "https://p.scdn.co/mp3-preview/1c0da00b5c95a1a6c9dfc05b14a1a628a6e0ad73?cid=159ac88b1c534ed7ae41602f1e558a49",
          "track_number" : 14,
          "type" : "track",
          "uri" : "spotify:track:0QZ5yyl6B6utIWkxeBDxQN"
        } ],
        "limit" : 20,
        "next" : "https://api.spotify.com/v1/search?query=the+night+we+met&type=track&offset=20&limit=20",
        "offset" : 0,
        "previous" : null,
        "total" : 204
      }
    }))

    div = await setupAndQuerySearchForm()
  })

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(div)
    api.searchSongs.restore()
  })

  it('should trigger a song search when the search button is clicked', () => {
    // Note how this _isn’t_ a snapshot test because we’re checking whether a function was called with
    // the right arguments.
    expect(api.searchSongs.firstCall.args[0]).toEqual({
    // expect(api.searchSongs.getCall(0)).toEqual({
      // rating: 'pg-13',
      q: 'hello' // Our test search term.
    })
  })

  //// LEFT OFF HERE!!!!!!!!!!!!!!

  it('should populate the song results container when search results arrive', () => {
    // Our mock search results yield one image, so we expect our results container to have one child.
    const searchResults = div.querySelector('div.SearchResults')
    expect(searchResults.children.length).toEqual(1)
  })
})

describe('failed API calls', () => {
  let div
  beforeEach(async () => {
    sinon.stub(api, 'searchSongs')
    api.searchSongs.returns(Promise.reject('Mock failure'))

    div = await setupAndQuerySearchForm()
  })

  afterEach(() => {
    ReactDOM.unmountComponentAtNode(div)
    api.searchSongs.restore()
  })

  it('should display an alert when the API call fails', () => {
    // The document should contain the error div.
    const searchError = div.querySelector('div.error')
    expect(searchError.textContent).toEqual('Sorry, but something went wrong.')
  })
})
