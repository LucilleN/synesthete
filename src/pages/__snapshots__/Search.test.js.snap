// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initial state should start with a disabled search button 1`] = `
<form
  className="SearchForm-searchForm-2"
  onSubmit={[Function]}
>
  <p
    className="MuiTypography-root SearchForm-title-1 MuiTypography-body1"
  >
    search
  </p>
  <div
    className="MuiGrid-root SearchForm-searchBar-3 MuiGrid-container MuiGrid-justify-xs-center"
  >
    <input
      className="SearchForm-searchField-4"
      name="query"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value="search for a song"
    />
    <button
      className="SearchForm-buttonDisabled-6"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        search tracks
      </p>
    </button>
  </div>
  <div
    className="MuiGrid-root SearchForm-resultsContainer-7 MuiGrid-container"
  >
    <div
      className="SearchResults-noResults-142"
    >
      <p
        className="MuiTypography-root SearchResults-message-143 MuiTypography-body1"
      >
        Search for a song by title to view results!
      </p>
    </div>
  </div>
</form>
`;

exports[`initial state should start with a search field with default text 1`] = `
<form
  className="SearchForm-searchForm-2"
  onSubmit={[Function]}
>
  <p
    className="MuiTypography-root SearchForm-title-1 MuiTypography-body1"
  >
    search
  </p>
  <div
    className="MuiGrid-root SearchForm-searchBar-3 MuiGrid-container MuiGrid-justify-xs-center"
  >
    <input
      className="SearchForm-searchField-4"
      name="query"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value="search for a song"
    />
    <button
      className="SearchForm-buttonDisabled-6"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        search tracks
      </p>
    </button>
  </div>
  <div
    className="MuiGrid-root SearchForm-resultsContainer-7 MuiGrid-container"
  >
    <div
      className="SearchResults-noResults-142"
    >
      <p
        className="MuiTypography-root SearchResults-message-143 MuiTypography-body1"
      >
        Search for a song by title to view results!
      </p>
    </div>
  </div>
</form>
`;

exports[`should start with a search field with default text 1`] = `
<form
  className="SearchForm-searchForm-2"
  onSubmit={[Function]}
>
  <p
    className="MuiTypography-root SearchForm-title-1 MuiTypography-body1"
  >
    search
  </p>
  <div
    className="MuiGrid-root SearchForm-searchBar-3 MuiGrid-container MuiGrid-justify-xs-center"
  >
    <input
      className="SearchForm-searchField-4"
      name="query"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      type="text"
      value="search for a song"
    />
    <button
      className="SearchForm-buttonDisabled-6"
      disabled={true}
      onClick={[Function]}
      type="submit"
    >
      <p
        className="MuiTypography-root MuiTypography-body1"
      >
        search tracks
      </p>
    </button>
  </div>
  <div
    className="MuiGrid-root SearchForm-resultsContainer-7 MuiGrid-container"
  >
    <div
      className="SearchResults-noResults-142"
    >
      <p
        className="MuiTypography-root SearchResults-message-143 MuiTypography-body1"
      >
        Search for a song by title to view results!
      </p>
    </div>
  </div>
</form>
`;
